File: Source/RTSPPlayerServer/RTSPPlayerServer/Serializers/Interprocess/InterprocessSerializer.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace RTSPPlayerServer.Serializers.Interprocess
{
    internal class InterprocessSerializer : IInterprocessSerializer
    {
        /// <summary>
        /// Serializes the interprocess frame into a string message.
        /// </summary>
        /// <param name="interprocessFrame">Interprocess frame.</param>
        /// <returns>Serialized message.</returns>
        public string Serialize(InterprocessFrame interprocessFrame)
        {
            return interprocessFrame?.ParameterDictionary?.Aggregate(string.Empty,
                (current, parameter) => current + " " + $"{parameter.Key}={parameter.Value}").ToLower() ?? string.Empty;
        }

        /// <summary>
        /// Deserializes the string message into an interprocess frame.
        /// </summary>
        /// <param name="message">Serialized message.</param>
        /// <returns>Interprocess frame.</returns>
        public InterprocessFrame Deserialize(string message)
        {
            return new InterprocessFrame {ParameterDictionary = ParseMessage(message)};
        }
        
        /// <summary>
        /// Parses the string message into the key-value parameter dictionary.
        /// </summary>
        /// <param name="message">Serialized message.</param>
        /// <returns>Parameter dictionary.</returns>
        private static Dictionary<string, string> ParseMessage(string message)
        {
            var dictionary = new Dictionary<string, string>();
            var pairs = message?.ToLower().Split(' ', StringSplitOptions.RemoveEmptyEntries);
            if (!pairs?.Any() ?? true) return dictionary;

            foreach (var pair in pairs)
            {
                var index = pair.IndexOf('=');
                if (index <= 0 || index >= pair.Length - 1) continue;

                var key = pair.Substring(0, index).Trim();
                var value = pair.Substring(index + 1).Trim();
                
                if (!string.IsNullOrEmpty(key))
                    dictionary.TryAdd(key, value);
            }

            return dictionary;
        }
    }
}

File: Source/RTSPPlayerClient/Common/Serialization/InterprocessSerializer.cpp
/// \file InterprocessSerializer.cpp
/// \brief Contains definitions of classes and functions for processing
/// interprocess data.
/// \bug No known bugs.

#include "InterprocessSerializer.hpp"

/// A namespace that contains common classes and functions for data
/// serialization.
namespace Common::Serialization {

    /// Serializes the interprocess frame into a byte array.
    /// \details Delegates serialization to an overloaded function.
    /// \param[in]  frame   Interprocess frame.
    /// \return Byte array.
    QByteArray InterprocessSerializer::serialize(
        const InterprocessFrame& frame) const {

        QByteArray array;
        serialize(frame, array);
        return array;
    }

    /// Serializes the interprocess frame into a byte array.
    /// \details Concatenates the key-value dictionary into a byte array.
    /// \param[in]  frame   Interprocess frame.
    /// \param[out] array   Byte array.
    void InterprocessSerializer::serialize(const InterprocessFrame& frame,
                                           QByteArray& array) const {

        auto begin = frame.parameterDictionary.cbegin();
        auto end = frame.parameterDictionary.cend();

        for (auto it = begin; it != end; ++it)
            array += it.key().toUtf8() + '=' + it.value().toUtf8() + ' ';

        array += '\n';
    }

    /// Deserializes the byte array into an interprocess frame.
    /// \details Delegates deserialization to an overloaded function.
    /// \param[in]  array   Byte array.
    /// \return Interprocess frame.
    InterprocessFrame InterprocessSerializer::deserialize(
        const QByteArray& array) {

        InterprocessFrame frame;
        deserialize(array, frame);
        return frame;
    }

    /// Deserializes the data buffer into an interprocess frame.
    /// \details Delegates deserialization to an overloaded function.
    /// \param[in]  data    Data buffer.
    /// \param[in]  size    Buffer size.
    /// \return Interprocess frame.
    InterprocessFrame InterprocessSerializer::deserialize(
        const char* data, int size) {

        InterprocessFrame frame;
        deserialize(data, size, frame);
        return frame;
    }

    /// Deserializes the byte array into an interprocess frame.
    /// \details Parses the byte array into a key-value dictionary.
    /// \param[in]  array   Byte array.
    /// \param[out] frame   Interprocess frame.
    void InterprocessSerializer::deserialize(const QByteArray& array,
                                             InterprocessFrame& frame) {

        auto pairs = array.toLower().split(' ');

        for (const auto& pair : pairs) {
            auto index = pair.indexOf('=');
            if (index <= 0 || index >= pair.size() - 1) continue;

            auto key = pair.left(index).trimmed();
            auto value = pair.mid(index + 1).trimmed();

            if (!key.isEmpty())
                frame.parameterDictionary.insert(key, value);
        }
    }

    /// Deserializes the data buffer into an interprocess frame.
    /// \details Delegates deserialization to an overloaded function.
    /// \param[in]  data    Data buffer.
    /// \param[in]  size    Buffer size.
    /// \param[out] frame   Interprocess frame.
    void InterprocessSerializer::deserialize(const char* data, int size,
                                             InterprocessFrame& frame) {

        deserialize(QByteArray::fromRawData(data, size), frame);
    }
}
